homeassistant/components/tuya/switch.py

```diff
@@ -105,23 +108,26 @@ def _setup_entities(hass, device_ids: list):
                     DPCODE_UV,
                     DPCODE_WET,
                 ]:
-                    entities.append(TuyaHaSwitch(device, device_manager, function))
+                    tuya_ha_switch = TuyaHaSwitch(device, device_manager, function)
                     # Main device switch is handled by the Fan object
             elif device.category == "cwysj":
                 if function in [DPCODE_FRESET, DPCODE_UV, DPCODE_PRESET, DPCODE_WRESET]:
-                    entities.append(TuyaHaSwitch(device, device_manager, function))
+                    tuya_ha_switch = TuyaHaSwitch(device, device_manager, function)
 
                 if function.startswith(DPCODE_SWITCH):
                     # Main device switch
-                    entities.append(TuyaHaSwitch(device, device_manager, function))
-                    continue
+                    tuya_ha_switch = TuyaHaSwitch(device, device_manager, function)
             else:
                 if function.startswith(DPCODE_START):
-                    entities.append(TuyaHaSwitch(device, device_manager, function))
-                    continue
+                    tuya_ha_switch = TuyaHaSwitch(device, device_manager, function)
                 if function.startswith(DPCODE_SWITCH):
-                    entities.append(TuyaHaSwitch(device, device_manager, function))
-                    continue
+                    tuya_ha_switch = TuyaHaSwitch(device, device_manager, function)
+
+            if tuya_ha_switch is not None:
+                entities.append(tuya_ha_switch)
+                hass.data[DOMAIN][TUYA_HA_DEVICES][
+                    tuya_ha_switch.tuya_device.id
+                ] = tuya_ha_switch
```

Comment:
Don't store the entity in this shared container. Just store the device id in a set before signaling the platforms to add new entities, when discovering a new device.